content/en/blog/_posts/2024-04-15-devsecops-in-action-k8s-secrets.md

```diff
@@ -0,0 +1,562 @@
+---
+layout: blog
+title: "DevSecOps in Action: Kubernetes Secrets"
+slug: devsecops-in-action-k8s-secrets
+date: 2024-04-15
+author: >
+  Evgenii Frikin
+---
+
+
+The Secret is one of the basic resources that exists in the Kubernetes API. This
+kind of resource is used for different purposes and despite the name Secret,
+sensitive data is insecure to store in it. Recently considerable attention has
+been paid to
+[security improvement](/docs/tasks/administer-cluster/encrypt-data/)
+for kind of Secret resource, but Kubernetes still doesn't include appropriate
+out-of-the-box because Kubernetes philosophy carries simplicity in order not to
+compromise it engineers have to make sure Kubernetes Secret are saved
+
+According to
+[Good practices for Kubernetes Secrets](/docs/concepts/security/secrets-good-practices/)
+there are several options on how to improve security of this kind of Secret:
+encrypt secrets in the etcd, restrict the access to them via RBAC, configure
+access to external Secret, etc. However, this partially solves the security issue
+because they are cluster-side solutions and they help to improve security of the
+cluster overall.
+
+In the modern world, unilateral solutions aren't effective enough and can hide
+security issues known as
+[Security through obscurity](https://en.m.wikipedia.org/wiki/Security_through_obscurity).
+To avoid it, it's necessary to integrate
+[DevSecOps methodology](https://www.redhat.com/en/topics/devops/what-is-devsecops)
+which implies integrated security as a shared responsibility throughout the
+entire IT lifecycle. In other words, security solutions must start with the
+development process and apply to all other levels correspondingly.
+
+According to
+[OWASP Secrets Management Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Secrets_Management_Cheat_Sheet.html#5-containers-orchestrators),
+it's recommended to use Injection of Secret via file and in-memory. Provision of
+sensitive data via environment variables isn't recommended, but at the same time
+it's recommended in The
+[Twelve Factors](https://12factor.net/config).
```

Comment:
Right. 12factor describes storing configuration as environment variables but it doesn't mention that storing sensitive data as environment variables is very dangerous. That's why I emphasize it here.

Reply:
I don't think that you need to mention 12factor here. You can just say that storing sensitive data in environment variables is a bad idea because <reason>